File: android/src/main/java/com/raserad/uxfeedback/UXFeedbackModule.kt
package com.raserad.uxfeedback

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.Arguments
import com.facebook.react.modules.core.DeviceEventManagerModule
import ru.uxfeedback.pub.sdk.UxFeedback
import ru.uxfeedback.pub.sdk.UxFbOnEventsListener
import ru.uxfeedback.pub.sdk.UxFbOnLogListener
import ru.uxfeedback.pub.sdk.UxFbColor
import ru.uxfeedback.pub.sdk.UxFbDimen
import ru.uxfeedback.pub.sdk.UxFbFont
import kotlin.Exception

class UXFeedbackModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext), UxFbOnEventsListener, UxFbOnLogListener {

    override fun getName() = "UXFeedbackModule"

    @ReactMethod
    fun setup(config: ReadableMap, promise: Promise) {
        UxFeedback.sdk?.campaignListener = this
        UxFeedback.sdk?.logListener = this
        val settings = config.getMap("settings")
        if (settings !== null) {
            setSettings(settings)
        }
        val theme = config.getMap("theme")
        if (theme !== null) {
            setTheme(theme)
        }
        val properties = config.getMap("properties")
        if (properties !== null) {
            setProperties(properties)
        }
    }

    @ReactMethod
    fun setSettings(settings: ReadableMap) {
        reactContext.runOnUiQueueThread {
            UxFeedback.sdk?.settings?.apply {
                apiUrlDedicated = try {
                    settings.getString("endpointURL")!!
                } catch (e: Exception) {
                    apiUrlDedicated
                }
                startGlobalDelayTimer = try {
                    settings.getDouble("globalDelayTimer").toInt()
                } catch (e: Exception) {
                    startGlobalDelayTimer
                }
                slideInUiBlocked = try {
                    settings.getBoolean("slideInUiBlocked")
                } catch (e: Exception) {
                    slideInUiBlocked
                }
                debugEnabled = try {
                    settings.getBoolean("debugEnabled")
                } catch (e: Exception) {
                    debugEnabled
                }
                fieldsEventEnabled = try {
                    settings.getBoolean("fieldsEventEnabled")
                } catch (e: Exception) {
                    fieldsEventEnabled
                }
                val slideInUiBlackout = settings.getMap("slideInUiBlackout")
                if (slideInUiBlackout != null) {
                    slideInUiBlackoutColor = try {
                        UxFbColor.fromHex(slideInUiBlackout.getString("color")!!)
                    } catch (e: Exception) {
                        slideInUiBlackoutColor
                    }
                    slideInUiBlackoutOpacity = try {
                        slideInUiBlackout.getDouble("opacity").toInt()
                    } catch (e: Exception) {
                        slideInUiBlackoutBlur
                    }
                    slideInUiBlackoutBlur = try {
                        slideInUiBlackout.getDouble("blur").toInt()
                    } catch (e: Exception) {
                        slideInUiBlackoutBlur
                    }
                }
                val popupUiBlackout = settings.getMap("popupUiBlackout")
                if (popupUiBlackout != null) {
                    popupUiBlackoutColor = try {
                        UxFbColor.fromHex(popupUiBlackout.getString("color")!!)
                    } catch (e: Exception) {
                        popupUiBlackoutColor
                    }
                    popupUiBlackoutOpacity = try {
                        popupUiBlackout.getDouble("opacity").toInt()
                    } catch (e: Exception) {
                        popupUiBlackoutOpacity
                    }
                    popupUiBlackoutBlur = try {
                        popupUiBlackout.getDouble("blur").toInt()
                    } catch (e: Exception) {
                        popupUiBlackoutBlur
                    }
                }
                retryCount = try {
                    settings.getDouble("retryCount").toInt()
                } catch (e: Exception) {
                    retryCount
                }
                retryTimeout = try {
                    settings.getDouble("retryTimeout").toInt()
                } catch (e: Exception) {
                    retryTimeout
                }
                socketTimeout = try {
                    settings.getDouble("socketTimeout").toInt()
                } catch (e: Exception) {
                    socketTimeout
                }
            }
        }
    }

    @ReactMethod
    fun startCampaign(eventName: String) {
        reactContext.runOnUiQueueThread {
            UxFeedback.sdk?.startCampaign(eventName)
        }
    }

    @ReactMethod
    fun stopCampaign() {
        reactContext.runOnUiQueueThread {
            UxFeedback.sdk?.stopCampaign()
        }
    }

    @ReactMethod
    fun setProperties(properties: ReadableMap) {
        reactContext.runOnUiQueueThread {
            UxFeedback.sdk?.properties.apply {
                properties.entryIterator.forEach {
                    this?.put(it.key, it.value.toString())
                }
            }
        }
    }

    @ReactMethod
    fun setTheme(theme: ReadableMap) {
        UxFeedback.sdk?.theme?.apply {
            bgColor = try {
                UxFbColor.fromHex(theme.getString("bgColor")!!)
            } catch (e: Exception) {
                bgColor
            }
            iconColor = try {
                UxFbColor.fromHex(theme.getString("iconColor")!!)
            } catch (e: Exception) {
                iconColor
            }
            text01Color = try {
                UxFbColor.fromHex(theme.getString("text01Color")!!)
            } catch (e: Exception) {
                text01Color
            }
            text02Color = try {
                UxFbColor.fromHex(theme.getString("text02Color")!!)
            } catch (e: Exception) {
                text02Color
            }
            text03Color = try {
                UxFbColor.fromHex(theme.getString("text03Color")!!)
            } catch (e: Exception) {
                text03Color
            }
            mainColor = try {
                UxFbColor.fromHex(theme.getString("mainColor")!!)
            } catch (e: Exception) {
                mainColor
            }
            errorColorPrimary = try {
                UxFbColor.fromHex(theme.getString("errorColorPrimary")!!)
            } catch (e: Exception) {
                errorColorPrimary
            }
            errorColorSecondary = try {
                UxFbColor.fromHex(theme.getString("errorColorSecondary")!!)
            } catch (e: Exception) {
                errorColorSecondary
            }
            btnBgColor = try {
                UxFbColor.fromHex(theme.getString("btnBgColor")!!)
            } catch (e: Exception) {
                btnBgColor
            }
            btnBgColorActive = try {
                UxFbColor.fromHex(theme.getString("btnBgColorActive")!!)
            } catch (e: Exception) {
                btnBgColorActive
            }
            btnTextColor = try {
                UxFbColor.fromHex(theme.getString("btnTextColor")!!)
            } catch (e: Exception) {
                btnTextColor
            }
            inputBgColor = try {
                UxFbColor.fromHex(theme.getString("inputBgColor")!!)
            } catch (e: Exception) {
                inputBgColor
            }
            inputBorderColor = try {
                UxFbColor.fromHex(theme.getString("inputBorderColor")!!)
            } catch (e: Exception) {
                inputBorderColor
            }
            controlBgColor = try {
                UxFbColor.fromHex(theme.getString("controlBgColor")!!)
            } catch (e: Exception) {
                controlBgColor
            }
            controlBgColorActive = try {
                UxFbColor.fromHex(theme.getString("controlBgColorActive")!!)
            } catch (e: Exception) {
                controlBgColorActive
            }
            controlIconColor = try {
                UxFbColor.fromHex(theme.getString("controlIconColor")!!)
            } catch (e: Exception) {
                controlIconColor
            }
            formBorderRadius = try {
                UxFbDimen.fromDp(theme.getDouble("formBorderRadius").toInt())
            } catch (e: Exception) {
                formBorderRadius
            }
            btnBorderRadius = try {
                UxFbDimen.fromDp(theme.getDouble("btnBorderRadius").toInt())
            } catch (e: Exception) {
                formBorderRadius
            }
            lightNavigationBar = try {
                theme.getBoolean("lightNavigationBar")
            } catch (e: Exception) {
                lightNavigationBar
            }
            fontBtn = getFontFromData(theme.getMap("fontBtn"), fontBtn)
            fontH1 = getFontFromData(theme.getMap("fontH1"), fontH1)
            fontH2 = getFontFromData(theme.getMap("fontH2"), fontH2)
            fontP1 = getFontFromData(theme.getMap("fontP1"), fontP1)
            fontP2 = getFontFromData(theme.getMap("fontP2"), fontP2)
        }
    }

    @ReactMethod
    fun addListener(eventName: String) {}

    @ReactMethod
    fun removeListeners(count: Int) {}

    override fun getConstants(): MutableMap<String, Any> {
        return hashMapOf("count" to 1)
    }

    private fun getFontFromData(data: ReadableMap?, defaultFont: UxFbFont): UxFbFont {
        if (data === null) {
            return defaultFont
        }
        return try {
            var family = data.getString("family")!!
            val fonts = currentActivity!!.resources.assets.list("fonts")!!
            if (fonts.contains("$family.otf")) {
                family = "$family.otf"
            } else if (fonts.contains("$family.ttf")) {
                family = "$family.ttf"
            } else if (fonts.contains("$family.woff")) {
                family = "$family.woff"
            }
            UxFbFont.fromAssets(
                    assetsPatch = "fonts/$family",
                    weight = data.getDouble("weight").toInt(),
                    sizeSp = data.getDouble("size").toInt(),
                    italic = data.getBoolean("italic")
            )
        } catch (e: Exception) {
            defaultFont
        }
    }

    private fun emitEvent(event: String, data: Any) {
        reactContext.runOnUiQueueThread {
            R.style.UXFBLightTheme
            reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java).emit(event, data)
        }
    }

    override fun uxFbNoCampaignToStart(eventName: String) {
        emitEvent("campaign_not_found", eventName)
    }

    override fun uxFbOnFieldsEvent(campaignId: Int, eventName: String, fieldValues: Map<String, Array<String>>) {
        val params = Arguments.createMap()
        params.putString("campaignId", campaignId.toString())
        params.putMap("fieldValues", Arguments.createMap().apply {
            fieldValues.forEach {
                this.putString(it.key, it.value.joinToString(","))
            }
        })
        emitEvent("campaign_event_send", params)
    }

    override fun uxFbOnFinishCampaign(campaignId: Int, eventName: String) {
        emitEvent("campaign_finish", eventName)
    }

    override fun uxFbOnReady() {
        emitEvent("campaign_loaded", true)
    }

    override fun uxFbOnStartCampaign(campaignId: Int, eventName: String) {
        emitEvent("campaign_show", eventName)
    }

    override fun uxFbOnTerminateCampaign(campaignId: Int, eventName: String, terminatedPage: Int, totalPages: Int) {
        val params = Arguments.createMap()
        params.putString("eventName", eventName)
        params.putInt("terminatePage", terminatedPage)
        params.putInt("totalPages", totalPages)
        emitEvent("campaign_terminate", params)
    }

    override fun uxFbOnLog(message: String) {
        emitEvent("log", message)
    }
}

File: ios/UXFeedbackModule.swift
import Foundation
import UXFeedbackSDK

@objc(UXFeedbackModule)
class UXFeedbackModule: RCTEventEmitter {
    
  var currentAppID = ""
  var currentSettings = UXFBSettings()

  @objc
  override func constantsToExport() -> [AnyHashable : Any]! {
    return [:]
  }

  @objc(setup:withResolver:withRejecter:)
  func setup(config: Dictionary<String, Any>, resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
      DispatchQueue.main.async {
          if let appID = config["iosAppID"] as? String {
              self.currentAppID = appID
              UXFeedback.setup(appID: appID, settings: self.currentSettings, campaignDelegate: self, logDelegate: self)
          }
          if let settings = config["settings"] as? Dictionary<String, Any> {
              self.setSettings(settings: settings)
          }
          if let theme = config["theme"] as? Dictionary<String, Any> {
              self.setTheme(theme: theme)
          }
          if let properties = config["properties"] as? Dictionary<String, Any> {
              self.setProperties(properties: properties)
          }
      }
  }

  @objc(setSettings:)
  func setSettings(settings: Dictionary<String, Any>) {
    DispatchQueue.main.async {
        let sdkSettings = UXFeedback.sdk.settings
        if let globalDelayTimer = settings["globalDelayTimer"] as? Int {
            sdkSettings.globalDelayTimer = globalDelayTimer
        }
        if let slideInUiBlocked = settings["slideInUiBlocked"] as? Bool {
            sdkSettings.slideInUiBlocked = slideInUiBlocked
        }
        if let debugEnabled = settings["debugEnabled"] as? Bool {
            sdkSettings.debugEnabled = debugEnabled
        }
        if let fieldsEventEnabled = settings["fieldsEventEnabled"] as? Bool {
            sdkSettings.fieldsEventEnabled = fieldsEventEnabled
        }
        if let retryTimout = settings["retryTimout"] as? Double {
            sdkSettings.retryTimeout = retryTimout
        }
        if let retryCount = settings["retryCount"] as? Int {
            sdkSettings.retryCount = retryCount
        }
        if let socketTimeout = settings["socketTimeout"] as? Double {
            sdkSettings.socketTimeout = socketTimeout
        }
        if let slideInUiBlackout = settings["slideInUiBlackout"] as? Dictionary<String, Any> {
            sdkSettings.slideInUiBlackoutColor = slideInUiBlackout["color"] as? String ?? sdkSettings.slideInUiBlackoutColor
            sdkSettings.slideInUiBlackoutBlur = slideInUiBlackout["blur"] as? Int ?? sdkSettings.slideInUiBlackoutBlur
            sdkSettings.slideInUiBlackoutOpacity = slideInUiBlackout["opacity"] as? Int ?? sdkSettings.slideInUiBlackoutOpacity
        }
        if let popupUiBlackout = settings["popupUiBlackout"] as? Dictionary<String, Any> {
            sdkSettings.popupUiBlackoutColor = popupUiBlackout["color"] as? String ?? sdkSettings.popupUiBlackoutColor
            sdkSettings.popupUiBlackoutBlur = popupUiBlackout["blur"] as? Int ?? sdkSettings.popupUiBlackoutBlur
            sdkSettings.popupUiBlackoutOpacity = popupUiBlackout["opacity"] as? Int ?? sdkSettings.popupUiBlackoutOpacity
        }
        guard let iosSettings = settings["ios"] as? Dictionary<String, Any> else {
          return
        }
        if let closeOnSwipe = iosSettings["closeOnSwipe"] as? Bool {
            sdkSettings.closeOnSwipe = closeOnSwipe
        }
    }
  }
    
  @objc(setTheme:)
  func setTheme(theme: Dictionary<String, Any>) {
    DispatchQueue.main.async {
        let sdkTheme = UXFeedback.sdk.theme
        
        if let text03Color = theme["text03Color"] as? String {
            sdkTheme.text03Color = UIColor.init(text03Color)
        }
        if let inputBorderColor = theme["inputBorderColor"] as? String {
            sdkTheme.inputBorderColor = UIColor.init(inputBorderColor)
        }
        if let iconColor = theme["iconColor"] as? String {
            sdkTheme.iconColor = UIColor.init(iconColor)
        }
        if let btnBgColorActive = theme["btnBgColorActive"] as? String {
            sdkTheme.btnBgColorActive = UIColor.init(btnBgColorActive)
        }
        if let btnBorderRadius = theme["btnBorderRadius"] as? Double {
            sdkTheme.btnBorderRadius = CGFloat(btnBorderRadius)
        }
        if let errorColorSecondary = theme["errorColorSecondary"] as? String {
            sdkTheme.errorColorSecondary = UIColor.init(errorColorSecondary)
        }
        if let errorColorPrimary = theme["errorColorPrimary"] as? String {
            sdkTheme.errorColorPrimary = UIColor.init(errorColorPrimary)
        }
        if let mainColor = theme["mainColor"] as? String {
            sdkTheme.mainColor = UIColor.init(mainColor)
        }
        if let controlBgColorActive = theme["controlBgColorActive"] as? String {
            sdkTheme.controlBgColorActive = UIColor.init(controlBgColorActive)
        }
        if let formBorderRadius = theme["formBorderRadius"] as? Double {
            sdkTheme.formBorderRadius = CGFloat(formBorderRadius)
        }
        if let inputBgColor = theme["inputBgColor"] as? String {
            sdkTheme.inputBgColor = UIColor.init(inputBgColor)
        }
        if let text01Color = theme["text01Color"] as? String {
            sdkTheme.text01Color = UIColor.init(text01Color)
        }
        if let controlBgColor = theme["controlBgColor"] as? String {
            sdkTheme.controlBgColor = UIColor.init(controlBgColor)
        }
        if let controlIconColor = theme["controlIconColor"] as? String {
            sdkTheme.controlIconColor = UIColor.init(controlIconColor)
        }
        if let btnBgColor = theme["btnBgColor"] as? String {
            sdkTheme.btnBgColor = UIColor.init(btnBgColor)
        }
        if let text02Color = theme["text02Color"] as? String {
            sdkTheme.text02Color = UIColor.init(text02Color)
        }
        if let btnTextColor = theme["btnTextColor"] as? String {
            sdkTheme.btnTextColor = UIColor.init(btnTextColor)
        }
        if let bgColor = theme["bgColor"] as? String {
            sdkTheme.bgColor = UIColor.init(bgColor)
        }
        if let fontH1 = theme["fontH1"] as? Dictionary<String, Any> {
            sdkTheme.fontH1 = self.getFont(fontH1, sdkTheme.fontH1)
        }
        if let fontH2 = theme["fontH2"] as? Dictionary<String, Any> {
            sdkTheme.fontH2 = self.getFont(fontH2, sdkTheme.fontH2)
        }
        if let fontP1 = theme["fontP1"] as? Dictionary<String, Any> {
            sdkTheme.fontP1 = self.getFont(fontP1, sdkTheme.fontP1)
        }
        if let fontP2 = theme["fontP2"] as? Dictionary<String, Any> {
            sdkTheme.fontP2 = self.getFont(fontP2, sdkTheme.fontP2)
        }
        if let fontBtn = theme["fontBtn"] as? Dictionary<String, Any> {
            sdkTheme.fontBtn = self.getFont(fontBtn, sdkTheme.fontBtn)
        }
    }
  }
    
    @objc(changeServer:)
    func changeServer(url: String) {
        DispatchQueue.main.async {
            let settings = self.currentSettings
            if (url == "") {
                settings.endpoint = nil
            } else {
                settings.endpoint = url
            }
            UXFeedback.setup(appID: self.currentAppID, settings: settings, campaignDelegate: self, logDelegate: self)
        }
    }

  @objc(startCampaign:)
  func startCampaign(eventName: String) {
    DispatchQueue.main.async {
        UXFeedback.sdk.startCampaign(eventName: eventName)
    }
  }

  @objc(stopCampaign)
  func stopCampaign() {
    DispatchQueue.main.async {
        UXFeedback.sdk.stopCampaign()
    }
  }

  @objc(setProperties:)
  func setProperties(properties: Dictionary<String, Any>) {
    DispatchQueue.main.async {
        properties.forEach { (key: String, value: Any) in
            UXFeedback.sdk.properties[key] = value
        }
    }
  }

  @objc
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
    
  @objc
  override func supportedEvents() -> [String]! {
    return ["campaign_show", "campaign_finish", "campaign_loaded", "campaign_event_send", "campaign_terminate", "log", "campaign_not_found"]
  }
    
  func emitEvent(withName: String, body: Any) {
    DispatchQueue.main.async {
        self.sendEvent(withName: withName, body: body)
    }
  }
}

extension UXFeedbackModule: UXFeedbackLogDelegate {
    func logDidReceive(message: String) {
        emitEvent(withName: "log", body: message)
    }
}

extension UXFeedbackModule: UXFeedbackCampaignDelegate {
    func campaignDidReceiveError(errorString: String) {
        emitEvent(withName: "campaign_not_found", body: errorString)
    }

    func campaignDidSend(campaignId: String) {}
    
    func campaignDidAnswered(campaignId: String, answers: [String : Any]) {
        emitEvent(withName: "campaign_event_send", body: ["campaignId": campaignId, "fieldValues": answers] as [String : Any])
    }

    func campaignDidTerminate(eventName: String, terminatedPage: Int, totalPages: Int) {
        emitEvent(withName: "campaign_terminate", body: ["eventName": eventName, "terminatePage": terminatedPage, "totalPages": totalPages] as [String : Any])
    }

    func campaignDidLoad(success: Bool) {
        emitEvent(withName: "campaign_loaded", body: success)
    }

    func campaignDidShow(eventName: String) {
        emitEvent(withName: "campaign_show", body: eventName)
    }

    func campaignDidClose(eventName: String) {
        emitEvent(withName: "campaign_finish", body: eventName)
    }
}

extension UXFeedbackModule {
    func getFontDescriptor(_ data: Dictionary<String, Any>, _ defaultFont: UIFont) -> UIFontDescriptor {
        return UIFontDescriptor(fontAttributes:
            [
                UIFontDescriptor.AttributeName.family: data["family"] as? String ?? defaultFont.familyName,
                UIFontDescriptor.AttributeName.traits: [
                    UIFontDescriptor.TraitKey.weight: self.getFontWeight(weight: data["weight"] as? String ?? ""),
                    UIFontDescriptor.TraitKey.slant: data["italic"] as? Bool ?? false ? 1.0 : 0.0
                ] as [UIFontDescriptor.TraitKey : Any]
            ]
        )
    }
    
    func getFont(_ data: Dictionary<String, Any>, _ defaultFont: UIFont) -> UIFont {
        return UIFont(
            descriptor: self.getFontDescriptor(data, defaultFont),
            size: CGFloat(data["size"] as? Int ?? Int(defaultFont.pointSize))
        )
    }
    
    func getFontWeight(weight: String) -> UIFont.Weight {
        switch (weight) {
            case "100": return .thin
            case "200": return .ultraLight
            case "300": return .light
            case "400": return .regular
            case "500": return .medium
            case "600": return .semibold
            case "700": return .bold
            case "800": return .heavy
            case "900": return .black
            default: return .regular
        }
    }
}

